### PR TITLE
Integrate encrypted logging

### LINKED ISSUE #3557: Implement Encrypted Logging
C6H8C3G23-p1612984324336700-slack

Encrypted Logging is already implemented in iOS. We can probably use the WPAndroid implementation for reference https://github.com/wordpress-mobile/WordPress-Android/pull/12571. 

Available for consult: @oguzkocer

### PR DESCRIPTION
**Pull request by wzieba**
Closes: #3557 

### Context

This PR integrates encrypted logging feature. You might be interested in `WPAndroid` PR , which did it earlier: https://github.com/wordpress-mobile/WordPress-Android/pull/12571/ - I've used some parts of code from there.

#### Preparations
1. Make sure repository with mobile secrets is up to date (`gradle.properties` has `wc.encrypted_logging_key` value)
2. You can use production Sentry DSN (no action required but you have to make sure to remove events after tests) or you can use some test project. If so, please update `wc.sentry.dsn` in `gradle.properties`.

#### TC1 - Upload encrypted logs with logging event
1. Apply `git patch` from the bottom of the PR description
2. Run the app
3. **Check Android logger prints `Encrypted log with uuid: <uuid> uploaded successfully!`**
4. Open Sentry console
5. Look for info-level event with message `This is just a test, ignore me`, open details
6. Scroll down to `Additional data`
7. **Assert there's `uuid` value and it's equal to `uuid` from point 3**
8. Go to `encrypted logs` feature in mc (`Tools` menu)
9. **Paste `uuid`, click `View` and assert that log is present**

#### TC2 - Upload encrypted logs on app start
1. Apply `git patch` from the bottom of the PR description
2. Turn off the wifi on test device
3. Run the app
4. **Check Android logger does not print `Encrypted log with uuid: <uuid> uploaded successfully!`**
5. Close the app
6. Turn on wifi
7. Run the app again
8. **Check Android logger prints 2 `Encrypted log with uuid: <uuid> uploaded successfully!`** events
9. Open `Sentry` console
10. Look for 2 info-level events with message `This is just a test, ignore me`, open details
11. Scroll down to `Additional data` on both
12. **Assert there's `uuid` value and it's equal to `uuid` one of the uudis from point 8. Do this for both events**
13. Go to `encrypted logs` feature in mc (`Tools` menu)
14. **Paste `uuid`, click `View` and assert that log is present. Do this for both events**

#### `git patch`

```
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/util/crashlogging/WCCrashLoggingDataProvider.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- WooCommerce/src/main/kotlin/com/woocommerce/android/util/crashlogging/WCCrashLoggingDataProvider.kt	(revision 8e9c52958466ec789570e9530503410d0b16777c)
+++ WooCommerce/src/main/kotlin/com/woocommerce/android/util/crashlogging/WCCrashLoggingDataProvider.kt	(date 1620298667000)
@@ -41,7 +41,7 @@
     }
 
     override fun crashLoggingEnabled(): Boolean {
-        return appPrefs.isCrashReportingEnabled()
+        return true
     }
 
     override fun extraKnownKeys(): List<ExtraKnownKey> {
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreFragment.kt	(revision 8e9c52958466ec789570e9530503410d0b16777c)
+++ WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreFragment.kt	(date 1620298667000)
@@ -181,6 +181,8 @@
         tabLayout.addOnTabSelectedListener(tabSelectedListener)
 
         refreshMyStoreStats(forced = this.isRefreshPending)
+
+        crashLogging.sendReport(message = "This is just a test, ignore me")
     }
 
     override fun onResume() {
```

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by AliSoftware** [comment]
@wzieba FYI I've moved milestone to 6.8 since 6.7 is being code-frozen today. Just wanted to be sure this Draft PR didn't fall into limbo 🙂

File: WooCommerce/build.gradle [hunk 1]
```diff
@@ -155,7 +155,7 @@ dependencies {
         exclude group: "com.android.support"
     }
 
-    implementation 'com.github.Automattic:Automattic-Tracks-Android:2.0.0'
+    implementation 'com.github.Automattic:Automattic-Tracks-Android:2.0.1'
```
Comment (wzieba) [inline WooCommerce/build.gradle:169]: Just a fix, release notes: https://github.com/Automattic/Automattic-Tracks-Android/releases/tag/2.0.1

File: WooCommerce/src/main/kotlin/com/woocommerce/android/tools/NetworkStatus.kt [hunk 2]
```diff
@@ -1,10 +1,10 @@
 package com.woocommerce.android.tools
 
 import android.content.Context
+import dagger.Reusable
 import org.wordpress.android.util.NetworkUtils
-import javax.inject.Singleton
 
-@Singleton
-class NetworkStatus(private var context: Context) {
+@Reusable
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/tools/NetworkStatus.kt:7]: This is not strictly connected to the encrypted logging. I've changed `NetworkStatus` from a singleton to a reusable class as it doesn't keep any state so there's no reason to add an additional constraint

File: WooCommerce/src/main/kotlin/com/woocommerce/android/util/WooLogWrapper.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.woocommerce.android.util
+
+import javax.inject.Inject
+
+class WooLogWrapper @Inject constructor() {
+    fun provideLogs(): String = WooLog.toString()
+}
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/util/WooLogWrapper.kt:7]: @malinajirka just a heads up that I'm proposing such a wrapper in this PR. I know you're planning to use `WooLog` in your PRs at some point so maybe you could add methods to this class if it'll suits your needs.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/util/crashlogging/EnqueueSendingEncryptedLogs.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+import com.automattic.android.tracks.crashlogging.EventLevel
+import com.automattic.android.tracks.crashlogging.EventLevel.FATAL
+import com.woocommerce.android.tools.NetworkStatus
+import org.wordpress.android.fluxc.Dispatcher
+import org.wordpress.android.fluxc.generated.EncryptedLogActionBuilder
+import org.wordpress.android.fluxc.store.EncryptedLogStore.UploadEncryptedLogPayload
+import javax.inject.Inject
+
+class EnqueueSendingEncryptedLogs @Inject constructor(
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/util/crashlogging/EnqueueSendingEncryptedLogs.kt:11]: Comparing to `WPAndroid` implementation, I've divided responsibilities of the [`EncryptedLogging`](https://github.com/wordpress-mobile/WordPress-Android/blob/70ce7de8817020d29768074bb408688436320220/WordPress/src/main/java/org/wordpress/android/util/EncryptedLogging.kt) class.

In `EnqueueSendingEncryptedLogs` we enqueue logs, in `UploadEncryptedLogs` we upload them and in `ObserveEncryptedLogsUploadResult` - observe results

File: WooCommerce/src/test/kotlin/com/woocommerce/android/util/crashlogging/WooLogFileProviderTest.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+    private val wooLogWrapper: WooLogWrapper = mock()
+
+    @Before
+    fun setUp() {
+        sut = WooLogFileProvider(wooLogWrapper)
+    }
+
+    @Test
+    fun `should provide a valid log file`() {
```
Comment (wzieba) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/util/crashlogging/WooLogFileProviderTest.kt:24]: One might think that such tests are overkill but did you know that `File#createTempFile`:
image
? 😄

File: gradle.properties-example [hunk 6]
```diff
@@ -32,6 +32,8 @@ wc.zendesk.domain = http://www.google.com/
 wc.zendesk.oauth_client_id = woocommerce
 wc.sentry.dsn = https://00000000000000000000000000000000@sentry.io/0000000
 
+wc.encrypted_logging_key=z0g+oVkqR4kWNUTxJfTozOZQjfXI7W9f6bD0uMJ5VkA=
```
Comment (wzieba) [inline gradle.properties-example:36]: Don't worry it's not a leak. This is [the same sample key](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/gradle.properties-example#L30) used in `WPAndroid` repository.

**Comment by wzieba** [comment]
Please don't bother with dependabot fail, [it's discussed](https://github.com/woocommerce/woocommerce-android/pull/3960#issuecomment-833330101).

File: gradle.properties-example (hunk 6 above)
Comment (AliSoftware) [reply gradle.properties-example:36]: worth a comment in the example file directly? 🤔

**Review by hichamboushaba — APPROVED** [review]
Great work @wzieba, I did several tests, and it works as expected, and hopefully it will help fixing crashes.

One question though, I see the initial implementation was using WorkManager, and then you dropped it: https://github.com/woocommerce/woocommerce-android/pull/3978/commits/8e9c52958466ec789570e9530503410d0b16777c, can I understand why? just to learn, I have nothing against the other approach

**Comment by wzieba** [comment]
@hichamboushaba thank you for the review!

True, I was testing various approaches and one of them was WorkManager. I've dropped it to reduce the complexity. Encrypted logs are meant to be sent in two cases: on app start and when a crash is reported (during app lifecycle). While I was proposing WorkManager, I thought that it might be a good idea to do this periodically but eventually, I've realised it's not a better solution.

As WorkManager didn't bring any value and moved us away from implementation in `WPAndroid`, I've decided to drop it.

File: gradle.properties-example (hunk 6 above)
Comment (wzieba) [reply gradle.properties-example:36]: Sure! Sorry, didn't spot it earlier, I've added a comment in https://github.com/woocommerce/woocommerce-android/pull/3978/commits/41a5a87e3cc970d3316a27212b19ad4914d70ed0

**Comment by wzieba** [comment]
@hichamboushaba just a heads up: since your review I've added a comment to `gradle.properties-example` and merged develop. As those are not any impactful changes, I'll merge the PR. 

Thank you again for the review!

**Pull request merged**